devicetypes/smartthings/rgbw-light.src/rgbw-light.groovy

```diff
@@ -232,53 +268,109 @@ def setHue(value) {
 
 def setColor(value) {
 	log.debug "setColor($value)"
-	def result = []
+	def rgb
+	if (state.staged == null) {
+		state.staged = [:]
+	}
 	if (value.hex) {
-		def c = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) }
-		result << zwave.switchColorV3.switchColorSet(red:c[0], green:c[1], blue:c[2], warmWhite:0, coldWhite:0)
+		state.staged << [color: value.hex] // stage ST RGB color attribute
+		def hsv = colorUtil.hexToHsv(value.hex) // convert to HSV
+		state.staged << [hue: hsv[0], saturation: hsv[1]] // stage ST hue and saturation attributes
+		rgb = value.hex.findAll(/[0-9a-fA-F]{2}/).collect { Integer.parseInt(it, 16) } // separate RGB elements for zwave setter
 	} else {
-		def rgb = huesatToRGB(value.hue, value.saturation)
-		result << zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2], warmWhite:0, coldWhite:0)
+		state.staged << value.subMap("hue", "saturation") // stage ST hue and saturation attributes
+		def hex = colorUtil.hsvToHex(Math.round(value.hue) as int, Math.round(value.saturation) as int) // convert to hex
+		state.staged << [color: hex] // stage ST RGB color attribute
+		rgb = colorUtil.hexToRgb(hex) // separate RGB elements for zwave setter
+	}
+	commands([zwave.switchColorV3.switchColorSet(red: rgb[0], green: rgb[1], blue: rgb[2], warmWhite: 0, coldWhite: 0),
+	          zwave.switchColorV3.switchColorGet(colorComponent: RGB_NAMES[0]), // event-publish callback is on any of the RGB responses, so only need to GET one of these
+	], 3500)
+}
+
+private emitTemperatureSet(temp, cmds) {
+	// Restrict temp to legal bounds.
+	temp = BOUND(temp, COLOR_TEMP_MIN, COLOR_TEMP_MAX)
+	// Add our staging dictionary to state.
+	if (state.staged == null) {
+		state.staged = [:]
 	}
-	if (device.currentValue("switch") != "on") {
-		result << zwave.basicV1.basicSet(value: 0xFF)
-		result << zwave.switchMultilevelV3.switchMultilevelGet()
+	// Stage ST colorTemperature attribute.
+	state.staged << [colorTemperature: temp]
+	// Emit our command.  Follow up with callback-trigger getter.  Our
+	// event-publish callback is on any of the while-level responses,
+	// so we only need to GET one of these these.  Make sure that we
+	// only have delay immediately before the callback trigger getter.
+	if (cmds.size() > 1) {
+		def prologue = cmds.init()
+		cmds = cmds.tail()
+		commands(prologue, 0) // emit in quick succession
```

Comment:
I'm not sure that I follow exactly what is going on here. Maybe I don't fully understand Groovy, but wouldn't the `commands` call here ultimately not get returned? `commands` ultimately calls `delayBetween` which returns an array or command strings that you can either return from a Capability command or pass to `sendHubEvent`. I think in this case the `commands(epilogue, 3500)` below will be the only thing returned. Unless Groovy has a feature where it takes both results from the calls to `commands` and adds them and returns that... (I don't admit to having a deep knowledge of Groovy's quirks.)

Reply:
I assumed that delayBetween was calling sendHubEvent, or something to this effect.  Certainly, that's the only way this _can_ work and I was sure it did.

I'll investigate tomorrow.

Does anybody on the review know what repo provides the DTH API?  I would like to see the delayBetween and sendHubEvent methods and am coming up short.

Reply:
@dkirker,

You are absolutely correct on this.  My code works, but in a backward unintended way resulting from my misunderstanding of the API.  Indeed, just calling commands and not ultimately returning it from the method doesn't do anything.

Fix will be forthcoming.

Reply:
Yup, you were right.  Fixed.  See update.